File: tests/integration/file_utils.py
#!/usr/bin/env python3
"""
File and git repository utilities for integration tests.

This module provides utilities for creating test environments,
managing git repositories, and safe directory cleanup.
"""

import contextlib
import os
import platform
import shutil
import subprocess
import tempfile
import time
from pathlib import Path


def create_test_environment() -> dict[str, str]:
    """
    Create a clean environment for testing.

    This removes CS_MOUNT_PATH and other variables that might interfere
    with testing static executable behavior.

    Returns:
        Environment dictionary suitable for subprocess execution
    """
    env = os.environ.copy()

    # Remove CS_MOUNT_PATH to test static mode
    env.pop("CS_MOUNT_PATH", None)

    return env


def create_git_repo(base_dir: Path, sample_files: dict[str, str]) -> Path:
    """
    Create a temporary git repository with sample files.

    Args:
        base_dir: Base directory where repo will be created
        sample_files: Dictionary mapping file paths to content

    Returns:
        Path to the created git repository
    """
    repo_dir = base_dir / "test_repo"
    repo_dir.mkdir(parents=True, exist_ok=True)

    # Initialize git repo
    subprocess.run(["git", "init", "-b", "master"], cwd=repo_dir, check=True, capture_output=True)
    subprocess.run(
        ["git", "config", "user.name", "Test User"],
        cwd=repo_dir,
        check=True,
        capture_output=True,
    )
    subprocess.run(
        ["git", "config", "user.email", "test@example.com"],
        cwd=repo_dir,
        check=True,
        capture_output=True,
    )

    # Create sample files
    for file_path, content in sample_files.items():
        full_path = repo_dir / file_path
        full_path.parent.mkdir(parents=True, exist_ok=True)
        full_path.write_text(content)

    # Initial commit
    subprocess.run(["git", "add", "."], cwd=repo_dir, check=True, capture_output=True)
    subprocess.run(
        ["git", "commit", "-m", "Initial commit"],
        cwd=repo_dir,
        check=True,
        capture_output=True,
    )

    return repo_dir


def _is_windows() -> bool:
    """Check if running on Windows."""
    return os.name == "nt" or platform.system() == "Windows"


def cleanup_dir(path: Path, retries: int = 5, delay: float = 0.5) -> None:
    """
    Safely remove a directory with retry logic for Windows.

    On Windows, file handles may be kept open by Git or other processes,
    causing PermissionError. This function retries with increasing delays.

    Args:
        path: Directory to remove
        retries: Number of retry attempts (default 5)
        delay: Initial delay between retries in seconds (default 0.5)
    """
    if not path.exists():
        return

    for attempt in range(retries):
        try:
            shutil.rmtree(path)
            return
        except PermissionError as e:
            if attempt < retries - 1:
                wait_time = delay * (attempt + 1)
                time.sleep(wait_time)
            else:
                print(f"  Warning: Failed to cleanup {path} after {retries} attempts: {e}")
        except Exception as e:
            print(f"  Warning: Failed to cleanup {path}: {e}")
            return


@contextlib.contextmanager
def safe_temp_directory(prefix: str = "cs_mcp_test_"):
    """
    Context manager for temporary directories with robust cleanup on Windows.

    Unlike tempfile.TemporaryDirectory, this handles Windows file locking
    by retrying cleanup with delays if files are still in use.

    Args:
        prefix: Prefix for the temporary directory name

    Yields:
        Path to the temporary directory
    """
    tmp_dir = Path(tempfile.mkdtemp(prefix=prefix)).resolve()
    try:
        yield tmp_dir
    finally:
        cleanup_dir(tmp_dir)

File: tests/e2e/file_utils.rs
//! File and git repository utilities for e2e tests.

use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;
use tempfile::TempDir;

/// Create a temporary git repository with sample files.
///
/// Returns the path to the repo directory within the temp dir.
pub fn create_git_repo(
    base_dir: &Path,
    sample_files: &HashMap<&str, &str>,
) -> Result<PathBuf, String> {
    let repo_dir = base_dir.join("test_repo");
    fs::create_dir_all(&repo_dir).map_err(|e| format!("Failed to create repo dir: {e}"))?;

    // Initialize git repo
    run_git(&repo_dir, &["init", "-b", "master"])?;
    run_git(&repo_dir, &["config", "user.name", "Test User"])?;
    run_git(&repo_dir, &["config", "user.email", "test@example.com"])?;
    run_git(&repo_dir, &["config", "index.version", "2"])?;

    // Create sample files
    for (file_path, content) in sample_files {
        let full_path = repo_dir.join(file_path);
        if let Some(parent) = full_path.parent() {
            fs::create_dir_all(parent)
                .map_err(|e| format!("Failed to create dir {}: {e}", parent.display()))?;
        }
        fs::write(&full_path, content)
            .map_err(|e| format!("Failed to write {}: {e}", full_path.display()))?;
    }

    // Initial commit
    run_git(&repo_dir, &["add", "."])?;
    run_git(&repo_dir, &["commit", "-m", "Initial commit"])?;

    Ok(repo_dir)
}

/// Run a git command in the given directory.
fn run_git(cwd: &Path, args: &[&str]) -> Result<(), String> {
    let output = Command::new("git")
        .args(args)
        .current_dir(cwd)
        .output()
        .map_err(|e| format!("Failed to run git {}: {e}", args.join(" ")))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("git {} failed: {stderr}", args.join(" ")));
    }
    Ok(())
}

/// Create a temporary directory that is automatically cleaned up on drop.
pub fn create_temp_dir(prefix: &str) -> Result<TempDir, String> {
    tempfile::Builder::new()
        .prefix(prefix)
        .tempdir()
        .map_err(|e| format!("Failed to create temp dir: {e}"))
}
